Sources/TuistDependencies/Carthage/CarthageInteractor.swift

```diff
@@ -0,0 +1,195 @@
+            try postInstallationActions(pathsProvider: pathsProvider)
+        }
+
+        switch method {
+        case .fetch: logger.info("Carthage dependencies were fetched successfully.", metadata: .success)
+        case .update: logger.info("Carthage dependencies were updated successfully.", metadata: .success)
+        }
+    }
+
+    // MARK: - Installation
+
+    private func prepareForInstallation(pathsProvider: PathsProvider, dependencies: [CarthageDependency]) throws {
+        // copy build directory from previous run if exist
+        if fileHandler.exists(pathsProvider.destinationCarthageDirectory) {
+            try copyDirectory(from: pathsProvider.destinationCarthageDirectory, to: pathsProvider.temporaryCarthageBuildDirectory)
+        }
+
+        // create `Cartfile`
+        let cartfileContent = cartfileContentGenerator.cartfileContent(for: dependencies)
+        let cartfilePath = pathsProvider.temporaryDirectoryPath.appending(component: "Cartfile")
+        try fileHandler.write(cartfileContent, path: cartfilePath, atomically: true)
+    }
+
+    private func postInstallationActions(pathsProvider: PathsProvider) throws {
+        // validation
+        guard fileHandler.exists(pathsProvider.temporaryCarfileResolvedPath) else {
+            throw CarthageInteractorError.cartfileNotFound
+        }
+        guard fileHandler.exists(pathsProvider.temporaryCarthageBuildDirectory) else {
+            throw CarthageInteractorError.buildDirectoryNotFound
+        }
+
+        // save `Cartfile.resolved`
+        try copyFile(from: pathsProvider.temporaryCarfileResolvedPath, to: pathsProvider.destinationCarfileResolvedPath)
+        // save build directory
+        try copyDirectory(from: pathsProvider.temporaryCarthageBuildDirectory, to: pathsProvider.destinationCarthageDirectory)
+    }
+
+    // MARK: - Helpers
+
+    private func copyFile(from fromPath: AbsolutePath, to toPath: AbsolutePath) throws {
+        try fileHandler.createFolder(toPath.removingLastComponent())
+
+        if fileHandler.exists(toPath) {
+            try fileHandler.replace(toPath, with: fromPath)
+        } else {
+            try fileHandler.copy(from: fromPath, to: toPath)
+        }
+    }
+
+    private func copyDirectory(from fromPath: AbsolutePath, to toPath: AbsolutePath) throws {
+        try fileHandler.createFolder(toPath.removingLastComponent())
+
+        if fileHandler.exists(toPath) {
+            try fileHandler.delete(toPath)
+        }
+
+        try fileHandler.copy(from: fromPath, to: toPath)
+    }
+
+    /// Returns true if Carthage is avaiable in the environment.
+    /// - Returns: True if Carthege is available globally in the system.
+    private func canUseSystemCarthage() -> Bool {
+        do {
+            _ = try System.shared.which("carthage")
+            return true
+        } catch {
+            return false
+        }
+    }
+}
+
+// MARK: - PathsProvider
+
+private extension CarthageInteractor {
```

Comment:
I don't think we need extensions here. We can rename `PathsProvider` to `CarthagePathsProvider`, make it `private`, and move it to the root of the file.

Reply:
I need it to prevent frameworks recompiling.